### ISSUE TITLE
Assert async-serialized content of dictionaries with ReferenceHandling.Preserve active

### ISSUE THREAD

**Issue by layomia** [dotnet/runtime#35611]
Following up on https://github.com/dotnet/runtime/pull/35046#discussion_r417063819:

In async tests for collections, we don't assert the result of serializing dictionaries with `ReferenceHandling.Preserve` active. This is because across multiple serialization operations, `ref` and `id` and metadata IDs may not be the same due to non-deterministic order of elements returned by `GetEnumerator()` for dictionaries. It is thus hard to compare the serialized results:

https://github.com/dotnet/runtime/blob/0fa0b905f65e57f5eb64dd556a07ead522930c8a/src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs#L98-L102

We can include these types by overriding the `GetEnumerator` method for these collections to return a deterministic order, and/or using types with deterministic order e.g. `OrderedDictionary`, `SortedDictionary`.

**Comment by JoshSchreuder** [comment]
@layomia I was able to add the deterministic dictionmary types pretty easily (I think), but what's the intended approach for `overriding the GetEnumerator method for these collections to return a deterministic order`?

Were you thinking something like subclassing all the dictionary types with overridden `GetEnumerator` and using those in the test, or is this possible to do with `Reflection.Emit` without using subclasses?

**Comment by layomia** [comment]
> Were you thinking something like subclassing all the dictionary types with overridden GetEnumerator and using those in the test

Yeah we can just use custom dictionaries in those tests. This would be a good approach if you're still interested in fixing this.

**Comment by layomia** [comment]
Moving to future as we have ample coverage for the relevant scenario, which is reference handling for POCOs.
